chandra_aca/centroid_resid.py

```diff
@@ -0,0 +1,325 @@
+                if fetch_source == 'cxc':
+                    if self.att_source == 'ground':
+                        self.centroid_dt = -2.4
+                    if self.att_source == 'obc':
+                        self.centroid_dt = -1.8
+                if fetch_source == 'maude':
+                    if self.att_source == 'ground':
+                        self.centroid_dt = -2.8
+                    if self.att_source == 'obc':
+                        self.centroid_dt = -2.34
+            else:
+                if fetch_source == 'cxc':
+                    if self.att_source == 'ground':
+                        self.centroid_dt = -2.44
+                    if self.att_source == 'obc':
+                        self.centroid_dt = -2.43
+                elif fetch_source == 'maude':
+                    if self.att_source == 'ground':
+                        self.centroid_dt = -2.45
+                    if self.att_source == 'obc':
+                        self.centroid_dt = -2.45
+
+        self.yag_times = self.yag_times + self.centroid_dt
+        self.zag_times = self.zag_times + self.centroid_dt
+        self._dt_applied = True
+
+    def calc_residuals(self):
+        """
+        Calculate residuals based on attitude and ra/dec of star.  Note that the sampling and times
+        of yags may be different from zags so these should be done independently.
+
+        Residuals are available in self.dyags and self.dzags.
+        Predicted values from attitude and star position in self.pred_yags and self.pred_zags
+
+        """
+        eci = Ska.quatutil.radec2eci(self.ra, self.dec)
+        # Transform the 3x3 to get the axes to align to have the dot product make sense
+        d_aca = np.dot(quat_vtransform(self.atts).transpose(0, 2, 1), eci)
+        p_yags = np.arctan2(d_aca[:, 1], d_aca[:, 0]) * R2A
+        p_zags = np.arctan2(d_aca[:, 2], d_aca[:, 0]) * R2A
+        self.pred_yags = interpolate(p_yags, self.att_times, self.yag_times, sorted=True)
+        self.pred_zags = interpolate(p_zags, self.att_times, self.zag_times, sorted=True)
+        self.dyags = self.yags - self.pred_yags
+        self.dzags = self.zags - self.pred_zags
+
+
+    @classmethod
+    def for_slot(cls, obsid=None, start=None, stop=None,
+                 slot=None, att_source='ground', centroid_source='ground'):
+        if obsid is not None:
+            if start is not None or stop is not None:
+                raise ValueError('cannot specify both obsid and start / stop')
+            ds = events.dwells.filter(obsid=obsid)
+            start = ds[0].start
+            stop = ds[len(ds) - 1].stop
+        if start is None or stop is None:
+            raise ValueError('must specify obsid or start / stop')
+        cr = cls(start, stop)
+        if obsid is not None:
+            cr.obsid = obsid
+        cr.set_atts(att_source)
+        cr.set_centroids(centroid_source, slot)
+        cr.set_star(obsid=obsid, slot=slot, date=start)
```

Comment:
Not quite getting it.  I'm guaranteed `date` so I never need obsid, and the methods to fetch the star catalog based on either are about the same speed so there's no optimization by providing `obsid`.